src/elements/common/current-user/withCurrentUser.tsx

```diff
@@ -0,0 +1,142 @@
+import React from 'react';
+// @ts-ignore: no ts definition
+import messages from '../messages';
+// @ts-ignore: no ts definition
+import { withAPIContext } from '../api-context';
+// @ts-ignore: no ts definition
+import { getBadItemError } from '../../../utils/error';
+// @ts-ignore: no ts definition
+// eslint-disable-next-line import/named
+import { ElementsXhrError, ErrorContextProps } from '../../../common/types/api';
+// @ts-ignore: no ts definition
+// eslint-disable-next-line import/named
+import { Errors } from '../flowTypes';
+// @ts-ignore: no ts definition
+// eslint-disable-next-line import/named
+import { User } from '../../../common/types/core';
+
+export type WithCurrentUserProps = {
+    currentUser?: User;
+    currentUserError?: Errors;
+} & ErrorContextProps;
+
+export type ComponentWithCurrentUser = {
+    errorCallback: (error: ElementsXhrError, code: string, contextInfo: Object) => void;
+    fetchCurrentUser: (user?: User, shouldDestroy?: boolean) => void;
+    fetchCurrentUserErrorCallback: (e: ElementsXhrError, code: string) => Errors;
+    fetchCurrentUserSuccessCallback: (currentUser: User) => void;
+};
+
+export type CurrentUserState = {
+    currentUser?: User;
+    currentUserError?: Errors;
+};
+
+export type WithCurrentUserComponent<P> = React.ComponentClass<P & WithCurrentUserProps>;
+
+export default function withCurrentUser<P extends object>(
+    WrappedComponent: React.ComponentType<P & WithCurrentUserProps>,
+): WithCurrentUserComponent<P> {
+    class ComponentWithCurrentUser extends React.Component<P & WithCurrentUserProps, CurrentUserState> {
```

Comment:
I assume we stuck with a class to reduce the effort involved in refactoring, but a function component would be nice someday. 😄

Reply:
To clarify: I don't think a conversion is needed as part of this PR, given the surface area of the change.